Day4

JustineSara opened the pull request.

JustineSara committed:
template for d4

JustineSara committed:
sample input

JustineSara committed:
all d4

gaverhae commented on src/t/d4.clj:
```diff
@@ -0,0 +1,164 @@
+  [input]
+  (->> input
+       cljstr/split-lines
+       (map-indexed (fn [i l]
+                      (->> l
+                      (map-indexed (fn [ii letter] {letter [[i ii]]}))
```
It's a minor thing but I found `i` and `ii` a bit hard to parse. Perhaps:
```suggestion
       (map-indexed (fn [y l]
                      (->> l
                      (map-indexed (fn [x letter] {letter [[y x]]}))
```
or `line` & `col`, but it's a bit awkward here with `lines` and `l `.

JustineSara replied:
I understand.
I won't change it though.
This i, ii, ... comes from a ... tweet or blog or something wondering why are we using i, then j, than k, and so on in loops, when you could have the level of your loop in a ore obvious way.

gaverhae replied:
That's actually a nice trick. 🤔

gaverhae commented on src/t/d4.clj:
```diff
@@ -0,0 +1,164 @@
+       (map-indexed (fn [i l]
+                      (->> l
+                      (map-indexed (fn [ii letter] {letter [[i ii]]}))
+                      (apply merge-with into))))
+       (apply merge-with into)))
```
This one is probably very subjective, but it took me a bit of time to figure out what `into` is doing here. Perhaps:
```suggestion
                      (apply merge-with (fnil conj [])))))
       (apply merge-with (fnil conj []))))
```
Though that requires changing the inner function to `{letter [i ii]}` instead.

I think overall I'd have written this as:

```clojure
(let [lines (cljstr/split-lines input)]
  (->> (for [y (range (count input))
            :let [line (nth lines y)] 
            x (range (count line))
            :let [c (nth line x)]]
         [c [y x]])
       (reduce (fn [acc [c p]] (update acc c (fnil conj #{}) p)))))
```
(untested), but I don't really have any concrete argument for whether that's better.

JustineSara replied:
this is because of the way I reach that point step by step, and the `(apply merge-with into)` is in the examples for `merge-with` so I used it, modified the input so it worked and moved on

gaverhae commented on src/t/d4.clj:
```diff
@@ -0,0 +1,164 @@
+       (apply merge-with into)))
+
+
+(comment
```
Any reason to keep it around?

JustineSara replied:
I worked hard for it, I want you to know

gaverhae commented on src/t/d4.clj:
```diff
@@ -0,0 +1,164 @@
+)
+
+(defn xmas?
+  [xpos dir Ms As Ss]
+  (let [x (first xpos)
+        y (second xpos)
+        dx (first dir)
+        dy (second dir)]
```
```suggestion
  [[x y] [dx dy] Ms As Ss]
```

JustineSara replied:
Ha yes, good point

gaverhae commented on src/t/d4.clj:
```diff
@@ -0,0 +1,164 @@
+      (contains? Ms [(+ x dx)       (+ y dy)])
+      (contains? As [(+ x dx dx)    (+ y dy dy)])
+      (contains? Ss [(+ x dx dx dx) (+ y dy dy dy)])
+      )))
```
```suggestion
      ))
```

gaverhae commented on src/t/d4.clj:
```diff
@@ -0,0 +1,164 @@
+        Ms (set (get p-input \M))
+        As (set (get p-input \A))
+        Ss (set (get p-input \S))
+        dirs (for [dx [+1 0 -1] dy [+1 0 -1]] [dx dy])
```
Don't you end up having `[0 0]` as a direction here? My gut instinct is to worry about that causing infinite loops somewhere.

JustineSara replied:
yes I have `[0 0]` in here, it's fine because I am testing if it is contained or not. It is not optimal because I am running a test I know will be negative (since I am testing for the next letter, not all 4 different letters can be at the same position) but it does not tremendously impact performance (at this point in Advent of Code at least)

JustineSara replied:
what would be the right way ?
```
:where (not (= dx dy 0))
```

gaverhae replied:
The form is correct I think (`(not (= dx dy 0))`), but it's `:when`, not `:where`. (I always have to look it up, that one somehow refuses to stick in my brain.)

gaverhae commented on src/t/d4.clj:
```diff
@@ -0,0 +1,164 @@
+        Ss (set (get p-input \S))
+        dirs (for [dx [+1 0 -1] dy [+1 0 -1]] [dx dy])
+        ]
+    (loop [x (first Xs)
+           otherXs (rest Xs)
+           tot 0]
+      (if (nil? x)
+        tot
+        (let [tots (map (fn [dir] (if (xmas? x dir Ms As Ss) 1 0)) dirs)]
+          (recur (first otherXs) (rest otherXs) (apply + tot tots)))
+      ))))
```
This looks like it could be a reduce:
```suggestion
    (reduce (fn [tot x] (+ tot (count (keep (fn [dir] (xmas? x dir Ms As Ss)) dirs))))
                  0
                  Xs)))
```

JustineSara replied:
oh yeah, that is probably right. I thought in term of loop and expected complex operation within ...

gaverhae reviewed (approved):
👍

JustineSara committed:
destructuring

(I realised later I needed it, was writing lots of `let` so got caught up in doing one more)

Co-authored-by: Gary Verhaegen <gary.verhaegen@gmail.com>

JustineSara committed:
destructuring - close parenthesis

Co-authored-by: Gary Verhaegen <gary.verhaegen@gmail.com>
